universal-application-tool-0.0.1/app/repository/ProgramRepository.java

```diff
@@ -33,16 +33,45 @@ public ProgramRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext execu
         executionContext);
   }
 
-  public CompletionStage<Void> insertProgram(Program program) {
+  public CompletionStage<Program> insertProgram(Program program) {
     return supplyAsync(
         () -> {
           ebeanServer.insert(program);
-          return null;
+          return program;
         },
         executionContext);
   }
 
-  public void insertProgramSync(Program program) {
+  public Program insertProgramSync(Program program) {
     ebeanServer.insert(program);
+    return program;
+  }
+
+  public CompletionStage<Program> updateProgram(Program program) {
+    return supplyAsync(
+        () -> {
+          ebeanServer.update(program);
+          return program;
+        },
+        executionContext);
+  }
+
+  public Program updateProgramSync(Program program) {
+    ebeanServer.update(program);
+    return program;
+  }
+
+  public CompletionStage<Program> insertOrUpdateProgram(Program program) {
```

Comment:
suggestion: I think the only path we need to have async is getting the program. All others are admin-only operations and can be synchronous

Reply:
Done!